src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender.cs

```diff
@@ -6,460 +6,542 @@ namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
     using Microsoft.VisualStudio.TestPlatform.ObjectModel;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
     using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
     using CommonResources = Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.Resources.Resources;
 
     /// <summary>
-    /// Utility class that facilitates the IPC comunication. Acts as server.
+    /// Test request sender implementation.
     /// </summary>
-    public sealed class TestRequestSender : ITestRequestSender
+    public class TestRequestSender : ITestRequestSender
     {
-        private ICommunicationManager communicationManager;
+        // Time to wait for test host exit (in seconds)
+        private const int CLIENTPROCESSEXITWAIT = 10 * 1000;
+
+        private readonly IDataSerializer dataSerializer;
+
+        private readonly ManualResetEventSlim connected;
+
+        private readonly ManualResetEventSlim clientExited;
+
+        private readonly int clientExitedWaitTime;
+
+        private ICommunicationEndPoint communicationEndpoint;
 
-        private ITransport transport;
+        private ICommunicationChannel channel;
 
-        private bool sendMessagesToRemoteHost = true;
+        private EventHandler<MessageReceivedEventArgs> onMessageReceived;
 
-        private IDataSerializer dataSerializer;
+        private Action<DisconnectedEventArgs> onDisconnected;
+
+        // Set to 1 if Discovery/Execution is complete, i.e. complete handlers have been invoked
+        private int operationCompleted;
+
+        private ITestMessageEventHandler messageEventHandler;
+
+        private string clientExitErrorMessage;
 
         // Set default to 1, if protocol version check does not happen
         // that implies host is using version 1
         private int protocolVersion = 1;
 
         private int highestSupportedVersion = 2;
-
-        /// <summary>
-        /// Use to cancel blocking tasks associated with testhost process
-        /// </summary>
-        private CancellationTokenSource clientExitCancellationSource;
-
-        private string clientExitErrorMessage;
+        private TestHostConnectionInfo connectionInfo;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="TestRequestSender"/> class.
         /// </summary>
-        /// <param name="protocolConfig">Protocol related information</param>
+        /// <param name="protocolConfig">Protocol configuration.</param>
         /// <param name="connectionInfo">Transport layer to set up connection</param>
         public TestRequestSender(ProtocolConfig protocolConfig, TestHostConnectionInfo connectionInfo)
-            : this(new SocketCommunicationManager(), connectionInfo, JsonDataSerializer.Instance, protocolConfig)
+            : this(connectionInfo, JsonDataSerializer.Instance, protocolConfig, CLIENTPROCESSEXITWAIT)
         {
+            if (this.connectionInfo.Role == ConnectionRole.Client)
```

Comment:
Please put a comment to move this to Communication factory, which returns ICommEndPoint based on Role, & ConnectionType (Socket, NamedPipe etc..) Once another communication mechanism is implemented